### PR TITLE
Allow 0 as "Repair bid" value

### PR DESCRIPTION
**Pull request by Falci**
Related to #163 

This changes depends on handshake-org/hsd/pull/477. 
Without the update on hsd, typing "0.000001" will show an error on each 0, removing the field focus (terrible UX);

### CODE REVIEW

**Comment by sdtsui** [comment]
@pinheadmz do you mind giving this a review?  Shoot me a DM if you want write access so that hsd/477 can be merged concurrently.

I suppose we'll also want a version bump in here?  If we do it this way, we'll want to make sure the lookahead var in our fork stays higher as some of our users have a lot of stored names.

File: app/pages/Auction/RepairBid.js
Comment (pinheadmz) [inline app/pages/Auction/RepairBid.js:7]: I don't think this is necessary, `showError` is connected a few lines down

File: app/pages/Auction/RepairBid.js
Comment (pinheadmz) [inline app/pages/Auction/RepairBid.js:65]: Mine as well move this to the top of the function I think.

**Comment by pinheadmz** [comment]
@Falci do you have the time to finish the other bugs in #163? If not I can add commits to this PR. I think we should nail it all in one move

File: app/pages/Auction/RepairBid.js
Comment (pinheadmz) [inline app/pages/Auction/RepairBid.js:60]: We still need a type check here. If the user enters for example `"foobar"` then `parseFloat()` returns `NaN` which will throw an error during `setState()`. Suggestion, maybe we need this:

```js
const dollarydoos = parseFloat(value);
if (!Number.isNaN(dollarydoos))
  return;
```

File: app/pages/Auction/RepairBid.js
Comment (Falci) [reply app/pages/Auction/RepairBid.js:7]: It seems like you forgot to import if before connecting. 
Without it, there is an console error, about `showError` not being defined.

File: app/pages/Auction/RepairBid.js
Comment (Falci) [reply app/pages/Auction/RepairBid.js:60]: `foobar` wouldn't pass the regex.

File: app/pages/Auction/RepairBid.js
Comment (Falci) [reply app/pages/Auction/RepairBid.js:65]: This needs to be after the `setState`. Otherwise, user can't clear the input.

Comment (pinheadmz) [reply app/pages/Auction/RepairBid.js:65]: great, thanks

File: app/pages/Auction/RepairBid.js
Comment (pinheadmz) [reply app/pages/Auction/RepairBid.js:7]: ok cool thank you

**Comment by Falci** [comment]
> @Falci do you have the time to finish the other bugs in #163? If not I can add commits to this PR. I think we should nail it all in one move

Do you mean enable 0 bid? I'm afraid I won't have time.

File: app/pages/Auction/RepairBid.js
Comment (pinheadmz) [inline app/pages/Auction/RepairBid.js:75]: Ok one more thing: we shouldn't change the value sent to `verifyBid()` because a few lines down `rpc importNonce` actually requires a float in HNS, not an integer in dollarydoos. The HTTP endpoint `getNonce` DOES accept an integer of dollarydoos so your intuition is correct, but this will break repairing any bid that is NOT `0`.

For this reason, I think we need to keep this line as `bid: value * consensus.COIN` and up on line 66, weneed to keep `return this.verifyBid(value);`

File: app/pages/Auction/RepairBid.js
Comment (pinheadmz) [inline app/pages/Auction/RepairBid.js:63]: I think `setState` needs to come after this check, so if the user tries to type `999999999999999`, it will just freeze once they get to MAX_MONEY, etc.

**Comment by pinheadmz** [comment]
https://github.com/handshake-org/hsd/pull/477 is merged, so this PR is now unblocked.

Optionally could be released with https://github.com/kyokan/bob-wallet/pull/170 which is blocked by https://github.com/handshake-org/hsd/pull/479

File: app/pages/Auction/RepairBid.js
Comment (Falci) [reply app/pages/Auction/RepairBid.js:63]: Nothing happens if you type a value too high.
The user doesn't know that we try to validate on each key pressed. In this case, we just don't try to validate.
It's harmless and transparent.

Comment (pinheadmz) [reply app/pages/Auction/RepairBid.js:63]: ok fair enough!

**Comment by Falci** [comment]
Do I need to do anything else to merge this?

**Comment by pinheadmz** [comment]
> Do I need to do anything else to merge this?

Just need to wait for repo maintainer review and approval :-) @mslipper @sdtsui @chikeichan

**Comment by chikeichan** [comment]
This PR looks good to me -- thanks @Falci !

@pinheadmz since Bob uses a forked version of `hsd` that uses a larger lookahead value, would this require us to update and cut a new version of `hsd`?

**Comment by chikeichan** [comment]
Used this branch for a couple of days and were able to verify the following:
- able to submit bid with local (http) custom RPC
- able to reveal bid with normal p2p
- able to send tx with remote (https) custom RPC

@pinheadmz just want to see if you were able to run this branch, and if you are okay with merging this 😃

**Comment by pinheadmz** [comment]
@chikeichan https://github.com/handshake-org/hsd/pull/477 is required for this patch to operate correctly in Bob. You could either cherry-pick that patch into the Kyokan hsd fork, or rebase your own patch set on hsd master before each Bob release.

To be honest, I think a PR to bump hsd lookahead to 1000 would probably be supported by the developers and get merged, even though it's a dumb "kick the can down the road" fix. But everyone with busy wallets is making this modification on their own anyway...

Related, allow Bob users to directly place 0-value bids: https://github.com/kyokan/bob-wallet/pull/170
This would require https://github.com/handshake-org/hsd/pull/460/files (already merged to hsd) and https://github.com/handshake-org/hsd/pull/479 (in review) to protect users from footgunning a dust output.

Probably https://github.com/handshake-org/hsd/pull/479 should be merged before next Bob release as well.

**Comment by chikeichan** [comment]
@pinheadmz thanks for the context! There is actually a [setLookahead](https://github.com/handshake-org/hsd/blob/master/lib/wallet/account.js#L566) method in the account object. We just need to query the object directly from `wdb` instead of getting it from wallet rpc, which is made a bit easier now that Bob is not using the wallet as a plugin. I just tested it and it allows one of my wallet with large lookahead to sync. Do you see potential problem with using this method?

It also seems like `hsd` has bumped `bcrypto` to match the [version](https://github.com/bcoin-org/bcrypto/32e6ce4c189e9c748373ce89bc6c7949f9a2992b) in our forked.

We likely don't need to maintain a fork anymore. I will file a issue to remind myself to verify and remove the fork when we get closer to next release.

**Comment by pinheadmz** [comment]
> @pinheadmz thanks for the context! There is actually a [setLookahead](https://github.com/handshake-org/hsd/blob/master/lib/wallet/account.js#L566) method in the account object. We just need to query the object directly from `wdb` instead of getting it from wallet rpc, which is made a bit easier now that Bob is not using the wallet as a plugin. I just tested it and it allows one of my wallet with large lookahead to sync. Do you see potential problem with using this method?

hm interesting - the `if (lookahead < this.lookahead)` condition looks a little funny to me in this block - like if the original value is `10` and you pass in `1` -- it will just add `9` to the `depth` values and return? 🤔 

So you're saying in Bob you just call this method "manually" and bump the lookahead without having to change the code in hsd? That actually looks ok -- and in fact you may only need to call that once after you create the account (the `"default"` account is created when you first create a wallet). The `lookahead` value is serialized and saved with the rest of the account metadata. 

The only issue I see with that is if someone uses unpatched hsd to restore the wallet. The default lookahead will be `10` again and they may not get a full BIP44 recovery. I think your method is ok for now but we should still look at bumping this value for all hsd.

**Comment by chikeichan** [comment]
yea it seems like the lookahead values also affects depths, so there is some special handling of the change in lookahead value to correctly adjust depth with the same change. i also verified you can only increase lookahead value, but not decrease it.

> The only issue I see with that is if someone uses unpatched hsd to restore the wallet. The default lookahead will be 10 again and they may not get a full BIP44 recovery. I think your method is ok for now but we should still look at bumping this value for all hsd.

almost seems like `hsd` would want this to be a config passed in thru CLI and the constructor as an option, or perhaps exposing this method thru RPC? (actually, exposing this method thru RPC seems pretty straightforward)

**Comment by pinheadmz** [comment]
The reason it's not an option or an RPC is because it is supposed to be a standard. [BIP44 actually defines the standard lookahead as 20](https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki#address-gap-limit). We don't really want users to have to remember an extra integer to correctly restore their wallet. Let's continue this here: https://github.com/handshake-org/hsd/issues/407

**Comment by Falci** [comment]
Hi,
There are 2 issues in #163: repair 0 bids and allow 0 bid.
Keep in mind this PR is just about the repair part.

**Review by chikeichan — APPROVED** [review]
this should fix #221

**Pull request merged**
